2주차/23827/23827_java_아키.java

```diff
@@ -0,0 +1,37 @@
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.math.BigInteger;
+import java.util.*;
+
+public class Main {
+
+    public static void main(String[] args) throws IOException {
+        //입력
+        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
+        int N = Integer.parseInt(br.readLine());
+        String[] inputs = br.readLine().split(" ");
+
+        BigInteger sum = new BigInteger("0");
+
+        long[] arr = new long[N];
+        for (int i = 0; i < N; i++) {
+            arr[i] = Integer.parseInt(inputs[i]);
+            BigInteger bigInteger = new BigInteger(String.valueOf(arr[i]));
+
+            sum = sum.add(bigInteger);
+        }
+
+        //알고리즘
+        BigInteger result = new BigInteger("0");
+        for (int i = 0; i < N - 1; i++) {
+            BigInteger bigInteger = new BigInteger(String.valueOf(arr[i]));
+            sum = sum.subtract(bigInteger);
+
+            result = result.add(bigInteger.multiply(sum));
+        }
```

Comment:
해당 문제의 경우 result에 `sum*arr[i]`의 값을 계속 더한다면 숫자가 overflow가 발생하여 BigInteger를 사용하신것 같은데 BigInteger는 사용 비용이 커서 가능하면 사용을 지양하는 편이 좋을 것 같아요.

이번 문제의 경우 아래와 같이 for문에서 나머지를 도출한다면 BigInteger없이 long으로도 해결 가능합니다😁
```suggestion
        for (int i = 0; i < N - 1; i++) {
            sum -= arr[i];
            result += (sum *  arr[i]);
            result %= 1_000_000_007;
        }
```

Reply:
제가 매번 결과값을 나누는 방법이 이해가 안되어서 BigInteger를 사용했었는데요,
케이가 답글로 알려주셨더라구요!
참고할 겸 여기에도 남깁니다

> 매 for문마다 나머지 연산을 진행해도 최종 결과는 변하지 않습니다. 
나머지 연산은 나머지를 제외한 사칙연산에 대해 아래 법칙이 성립합니다.

`(a+b+c)%M = (a%M + b%M + c%M) % M`
`(a*b*c)%M = (a%M * b%M * c%M) % M`